sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/OffsetRangeTracker.java

```diff
@@ -49,13 +50,22 @@ public OffsetRange currentRestriction() {
 
   @Override
   public SplitResult<OffsetRange> trySplit(double fractionOfRemainder) {
-    // TODO(BEAM-8872): Add support for splitting off a fixed amount of work for this restriction
-    // instead of only supporting checkpointing.
-
-    checkState(
-        lastClaimedOffset != null, "Can't checkpoint before any offset was successfully claimed");
-    OffsetRange res = new OffsetRange(lastClaimedOffset + 1, range.getTo());
-    this.range = new OffsetRange(range.getFrom(), lastClaimedOffset + 1);
+    checkState(lastClaimedOffset != null, "Can't split before any offset was successfully claimed");
+    // No more split should be performed if checkpoint has happened.
+    if (checkpointed) {
+      return null;
+    }
+    Long splitPos =
+        lastClaimedOffset
+            + Math.max(1L, (long) ((range.getTo() - lastClaimedOffset) * fractionOfRemainder));
+    if (splitPos >= range.getTo()) {
+      return null;
+    }
+    if (fractionOfRemainder == 0.0) {
+      checkpointed = true;
```

Comment:
Why do we need `checkpointed`?

Shouldn't the range restriction change so that `to` becomes `lastClaimed` (or `from` if nothing has been claimed)?

Reply:
Just return early since we know there is no more split after checkpointing.

Reply:
Changing the state makes the code more complicated though since the bounds checking varies based upon whether you got 0.0 or 0.00001 which is likely to produce the same final state with one having an additional boolean being set.